beacon/light/chain.go

```diff
@@ -159,3 +164,114 @@ func (c *LightClient) getExecutableData(head common.Hash) (*engine.ExecutableDat
+// versionedBlockToExecutableData parses versioned blocks and returns a generic
+// execution payload object.
+func versionedBlockToExecutableData(block *eth2spec.VersionedSignedBeaconBlock) *engine.ExecutableData {
+	var ep *engine.ExecutableData
+	switch block.Version {
+	case eth2spec.DataVersionPhase0:
+		panic("phase0 block has no execution payload to send")
+	case eth2spec.DataVersionAltair:
+		panic("altair block has no execution payload to send")
+	case eth2spec.DataVersionBellatrix:
+		p := block.Bellatrix.Message.Body.ExecutionPayload
+		ep = &engine.ExecutableData{
+			ParentHash:    common.Hash(p.ParentHash),
+			FeeRecipient:  common.Address(p.FeeRecipient),
+			StateRoot:     p.StateRoot,
+			ReceiptsRoot:  p.ReceiptsRoot,
+			LogsBloom:     p.LogsBloom[:],
+			Random:        p.PrevRandao,
+			Number:        p.BlockNumber,
+			GasLimit:      p.GasLimit,
+			GasUsed:       p.GasUsed,
+			Timestamp:     p.Timestamp,
+			ExtraData:     p.ExtraData,
+			BaseFeePerGas: new(big.Int).SetBytes(reverse(p.BaseFeePerGas[:])),
+			BlockHash:     common.Hash(p.BlockHash),
+			Transactions:  [][]byte{},
+			Withdrawals:   nil,
+			BlobGasUsed:   nil,
+			ExcessBlobGas: nil,
+		}
+		for _, tx := range p.Transactions {
+			ep.Transactions = append(ep.Transactions, tx)
+		}
+	case eth2spec.DataVersionCapella:
+		p := block.Capella.Message.Body.ExecutionPayload
+		ep = &engine.ExecutableData{
+			ParentHash:    common.Hash(p.ParentHash),
+			FeeRecipient:  common.Address(p.FeeRecipient),
+			StateRoot:     p.StateRoot,
+			ReceiptsRoot:  p.ReceiptsRoot,
+			LogsBloom:     p.LogsBloom[:],
+			Random:        p.PrevRandao,
+			Number:        p.BlockNumber,
+			GasLimit:      p.GasLimit,
+			GasUsed:       p.GasUsed,
+			Timestamp:     p.Timestamp,
+			ExtraData:     p.ExtraData,
+			BaseFeePerGas: new(big.Int).SetBytes(reverse(p.BaseFeePerGas[:])),
+			BlockHash:     common.Hash(p.BlockHash),
+			Transactions:  [][]byte{},
+			Withdrawals:   nil,
+			BlobGasUsed:   nil,
+			ExcessBlobGas: nil,
+		}
+		for _, tx := range p.Transactions {
+			ep.Transactions = append(ep.Transactions, tx)
+		}
+		for _, wx := range p.Withdrawals {
+			ep.Withdrawals = append(ep.Withdrawals, &ctypes.Withdrawal{
+				Index:     uint64(wx.Index),
+				Validator: uint64(wx.ValidatorIndex),
+				Address:   common.Address(wx.Address),
+				Amount:    uint64(wx.Amount),
+			})
+		}
+	case eth2spec.DataVersionDeneb:
+		p := block.Deneb.Message.Body.ExecutionPayload
+		ep = &engine.ExecutableData{
+			ParentHash:    common.Hash(p.ParentHash),
+			FeeRecipient:  common.Address(p.FeeRecipient),
+			StateRoot:     common.Hash(p.StateRoot),
+			ReceiptsRoot:  common.Hash(p.ReceiptsRoot),
+			LogsBloom:     p.LogsBloom[:],
+			Random:        p.PrevRandao,
+			Number:        p.BlockNumber,
+			GasLimit:      p.GasLimit,
+			GasUsed:       p.GasUsed,
+			Timestamp:     p.Timestamp,
+			ExtraData:     p.ExtraData,
+			BaseFeePerGas: nil, // TODO: convert this []uint64 correctly to big.Int
```

Comment:
I think you correctly convert this in the other cases

Reply:
nvm, just saw that this is just moved

Reply:
i'm not exactly sure when I put this comment here, but it seems like the deneb rendition from the eth2 lib has `BaseFeePerGas` as a `uint256` so I can just convert that to `big.Int`.